packages/libsql/src/lib/Category.ts

```diff
@@ -0,0 +1,417 @@
+      case "Written": {
+        const token = Token.create(result.position)
+        return { type: "Written", token }
+      }
+    }
+  }
+
+  static create({ client, readClient, batchSize, maxBatches }: ContextConfig) {
+    const connection = LibSqlConnection.create(client, readClient)
+    return new LibSqlContext(connection, batchSize, maxBatches)
+  }
+}
+
+export type AccessStrategy<Event, State> =
+  | { type: "Unoptimized" }
+  | { type: "LatestKnownEvent" }
+  | { type: "Snapshot"; isOrigin: (e: Event) => boolean; toSnapshot: (state: State) => Event }
+  | { type: "RollingState"; toSnapshot: (state: State) => Event }
+
+// prettier-ignore
+export namespace AccessStrategy {
+  export const Unoptimized = <E=any, S=any>(): AccessStrategy<E, S> => ({ type: "Unoptimized" })
+  export const LatestKnownEvent = <E=any, S=any>(): AccessStrategy<E, S> => ({ type: "LatestKnownEvent" })
+  export const Snapshot = <E, S>(isOrigin: (e: E) => boolean, toSnapshot: (state: S) => E): AccessStrategy<E, S> => 
+    ({ type: "Snapshot", isOrigin, toSnapshot })
+  export const RollingState = <E, S>(toSnapshot: (state: S) => E): AccessStrategy<E, S> => ({ type: "RollingState", toSnapshot })
+}
+
+class InternalCategory<Event, State, Context>
+  implements IReloadableCategory<Event, State, Context>
+{
+  constructor(
+    private readonly context: LibSqlContext,
+    private readonly categoryName: string,
+    private readonly codec: ICodec<Event, Format, Context>,
+    private readonly fold: (state: State, events: Event[]) => State,
+    private readonly initial: State,
+    private readonly access: AccessStrategy<Event, State> = AccessStrategy.Unoptimized(),
+  ) {}
+
+  supersedes = Token.supersedes
+
+  // prettier-ignore
+  private async loadAlgorithm(streamId: Equinox.StreamId): Promise<[StreamToken, State]> {
+    const streamName = Equinox.StreamName.create(this.categoryName, streamId)
+    const span = trace.getActiveSpan()
+    span?.setAttributes({
+      [Tags.access_strategy]: this.access.type,
+      [Tags.category]: this.categoryName,
+      [Tags.stream_name]: streamName,
+    })
+    switch (this.access.type) {
+      case "Unoptimized":
+        return this.context.loadBatched(streamName, this.codec.decode, this.fold, this.initial)
+      case "LatestKnownEvent":
+        return this.context.loadLast(streamName, this.codec.decode, this.fold, this.initial)
+      case "Snapshot":
+        return this.context.loadSnapshot(streamName, this.codec.decode, this.fold, this.initial, this.access.isOrigin)
+      case "RollingState":
+        return this.context.loadState(streamName, this.codec.decode, this.fold, this.initial)
+    }
+  }
+
+  async load(streamId: Equinox.StreamId, _maxStaleMs: number) {
+    const [token, state] = await this.loadAlgorithm(streamId)
+    return { token, state }
+  }
+
+  async reload(streamId: Equinox.StreamId, _requireLeader: boolean, t: TokenAndState<State>) {
+    if (this.access.type === "RollingState") {
```

Comment:
re earlier comment, for snapshot mode, cheapest reload is prob from that
maybe this is covered but naming and/or comments are not making it unambiguous in this fn

Reply:
You're right, I'm only optimising RollingState here. Snapshot reload behaviour is the same as MessageDB

Reply:
Will do something more comprehensive later